flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySliceFunction.java

```diff
@@ -0,0 +1,99 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_SLICE}. */
+@Internal
+public class ArraySliceFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+
+    public ArraySliceFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_SLICE, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(dataType.getLogicalType());
+    }
+
+    public @Nullable ArrayData eval(ArrayData array, Integer start, Integer end) {
+        try {
+            if (array == null || start == null || end == null) {
+                return null;
+            }
+            if (array.size() == 0) {
+                return array;
+            }
+
+            if (start == 0) {
+                start = 1;
+            }
+            if (end == 0) {
+                end = 1;
+            }
+
+            int startIndex =
+                    start < 0
+                            ? (start < -array.size() ? start : array.size() + start)
+                            : (start > array.size() ? start : start - 1);
+            int endIndex =
+                    end < 0
+                            ? (end < -array.size() ? end : array.size() + end)
+                            : (end > array.size() ? end : end - 1);
+
+            if (startIndex > endIndex) {
+                return new GenericArrayData(new Object[0]);
+            }
+            if (startIndex > array.size()) {
+                return new GenericArrayData(new Object[0]);
+            }
+            if (endIndex < -array.size()) {
+                return new GenericArrayData(new Object[0]);
+            }
+            if (endIndex > array.size()) {
+                endIndex = array.size() - 1;
+            }
+            if (startIndex < -array.size()) {
+                startIndex = 0;
+            }
+
+            List<Object> slicedArray = new ArrayList<>();
+
+            for (int i = startIndex; i <= endIndex; i++) {
+                slicedArray.add(elementGetter.getElementOrNull(array, i));
+            }
```

Comment:
```suggestion
            int startIndex = start;
            int endIndex = end;
            startIndex += startIndex < 0 ? array.size() + 1 : 0;
            endIndex += endIndex < 0 ? array.size() + 1 : 0;
            startIndex = Math.max(1, startIndex);
            endIndex = endIndex == 0 ? 1 : Math.min(endIndex, array.size());

            List<Object> slicedArray = new ArrayList<>();

            for (int i = startIndex - 1; i <= endIndex - 1; i++) {
                slicedArray.add(elementGetter.getElementOrNull(array, i));
            }
```

I think all this code could be simplified like that

Reply:
ok